File: Medium-Array/CountSubarray.java
import java.util.Scanner;

public class CountSubarray
{
    // Method to count subarrays whose sum is equal to k
    public int cntSubarrays(int[] arr, int k)
    {
        int count = 0;          // Initialize counter for subarrays
        int n = arr.length;     // Get the size of the array

        // Outer loop to fix the starting point of subarray
        for (int i = 0; i < n; i++)
        {
            int sum = 0;        // Initialize sum for the current subarray

            // Inner loop to extend the subarray from index i to j
            for (int j = i; j < n; j++)
            {
                sum += arr[j];  // Add current element to sum

                // Check if the sum equals k
                if (sum == k)
                {
                    count++;    // Increment count if sum matches k
                }
            }
        }

        return count;           // Return total count of subarrays
    }

    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);  // Scanner to take user input

        // Input array size
        System.out.print("Enter the size of the array: ");
        int n = sc.nextInt();
        int[] arr = new int[n];

        // Input array elements
        System.out.println("Enter the elements of the array:");
        for (int i = 0; i < n; i++)
        {
            arr[i] = sc.nextInt();
        }

        // Input target sum k
        System.out.print("Enter the value of k: ");
        int k = sc.nextInt();

        // Create object and call method to count subarrays
        CountSubarray cs = new CountSubarray();
        int result = cs.cntSubarrays(arr, k);

        // Output the result
        System.out.println("Number of subarrays with sum " + k + " is: " + result);

        sc.close();  // Close scanner
    }
}

File: Medium-Array/CountSubarray.py
# Problem Statement: Given an array of integers and an integer k, return the total number of subarrays whose sum equals k. A subarray is a contiguous non-empty sequence of elements within an array.

def count_subarrays_with_sum_k(arr, k):
    # count of subarrays with sum k
    count = 0
    # dictionary to store the cumulative sum frequencies
    cumulative_sum_freq = {0: 1}  # Initialize with sum 0 occurring
    cumulative_sum = 0
    for num in arr:
        cumulative_sum += num
        # Check if there is a cumulative sum that when subtracted from current sum gives k
        if (cumulative_sum - k) in cumulative_sum_freq:
            count += cumulative_sum_freq[cumulative_sum - k]
        # Update the frequency of the current cumulative sum
        cumulative_sum_freq[cumulative_sum] = cumulative_sum_freq.get(cumulative_sum, 0) + 1
    return count






nums=int(input("Enter the number of elements in the array: "))
print("Enter the elements of the array:")
arr = list(map(int, input().split()))
k=int(input("Enter the value of k: "))
result = count_subarrays_with_sum_k(arr, k)
print("The total number of subarrays whose sum equals k is:", result)